### PR TITLE
Fix (link): Don't undo autolinking when any other change has been made.

### LINKED ISSUE #11040: Auto-link correction
## 📝 Provide detailed reproduction steps (if any)

1. …type www.google.com in content (without http)
2. …inspect the link, you'll find it with href='www.google.com' which is invalid

### ✔️ Expected result

href='https://www.google.com'

### ❌ Actual result

href='www.google.com'

### ❓ Possible solution

Either auto-link would add https:// to the link, or doesn't link www.google.com

### LINKED ISSUE #12447: Auto-link undo correction
## 📝 Provide detailed reproduction steps (if any)

1. type "https://ckeditor.com abc" in content (hyperlink should be created for ckeditor.com, while "abc" still remains as text, which is fine)
2. slowly (one character at a time) hit backspace to remove " abc" text (including space).
3. Instead of removing space, editor undo character removal to initial "https://ckeditor.com abc" situation.

### ✔️ Expected result

just "https://ckeditor.com" in editor

### ❌ Actual result

"https://ckeditor.com abc" in editor

### PR DESCRIPTION
**Pull request by kbinieda**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Fix (link): Don't undo autolinking when any other change has been made. Closes #12447 .

### CODE REVIEW

**Comment by scofalik** [comment]
As far as I can see, the fix here is to prevent additional auto-linking of already linked text. AFAICS, This can happen in two scenarios:

* when you move selection before the link and type space again,
* when you start removing text, e.g. `http://foo.pl abc[]` -> `http://foo.pl []`.

I am quite sure that we don't want to double auto-link in such cases :).

However, I have two thoughts regarding this.

One, I wonder if `TextWatcher` should fire an event on text deletion. IDK if there are cases where this could be useful/used. In general, if you are deleting stuff, you must have added it earlier, which means that the `TextWatcher` already had a chance to act. Unless you pasted some text (e.g. you pasted `lorem http://foo.pl ipsum`). But in this case, AFAIK, we don't want `TextWatcher` to autolink after removing `ipsum` right? So, maybe we should also change how `TextWatcher` works? Nevertheless, the proposed if will help in the first case (moving selection before the link and typing space again).

Second, I think that I'd add one more unit test. Proposed unit test is fine, as it targets exactly the scenario described in the ticket. However, I'd add one more unit test, that would make sure that we don't autolink already linked text (you can use the scenario with moving selection and typing space again). This test will target the actual change.

cc @Reinmar (regarding `TextWatcher`).

**Review by scofalik — CHANGES_REQUESTED** [review]
Please add one more test (as described in my previous comment).

**Comment by kbinieda** [comment]
Aside from adding additional test, I've broke down _applyAutoLink function so that the moment of applying linkHref can be easily tested.

File: packages/ckeditor5-link/tests/autolink.js [hunk 1]
```diff
@@ -414,6 +425,27 @@ describe( 'AutoLink', () => {
 				'<paragraph>https://www.cksource.com []</paragraph>'
 			);
 		} );
+
+		it( 'should not undo auto-linking by pressing backspace after any other change has been made', () => {
```
Comment (scofalik) [inline packages/ckeditor5-link/tests/autolink.js:429]: Could you add a link to the issue in the comment above this line? This test will seem very weird to someone looking at it in a year :).

**Review by scofalik — CHANGES_REQUESTED** [review]
Small fixes required.

File: packages/ckeditor5-link/tests/autolink.js [hunk 2]
```diff
@@ -246,6 +246,17 @@ describe( 'AutoLink', () => {
 			);
 		} );
 
+		it( 'does not autolink if link is already created', () => {
```
[the remark this answers is not in the document]
Comment (kbinieda) [reply packages/ckeditor5-link/tests/autolink.js:249]: That's a good point. The "different text" approach gives us different result which we can then easily compare, I haven't thought of that. On the other hand If we want to check if same text/link hasn't been applied twice I would go for a method spy as it is more readable than operations counting.

**Comment by Reinmar** [comment]
> One, I wonder if `TextWatcher` should fire an event on text deletion. IDK if there are cases where this could be useful/is already used. In general, if you are deleting stuff, you must have added it earlier, which means that the `TextWatcher` already had a chance to act. Unless you pasted some text (e.g. you pasted `lorem http://foo.pl ipsum`). But in this case, AFAIK, we don't want `TextWatcher` to autolink after removing `ipsum` right? So, maybe we should also change how `TextWatcher` works?

TextWatcher is also used by mentions. I tested quickly how mentions work e.g. in Slack and when you start with this:

```
Hello, @reinmarX[]
```

And press delete, the mention panel will be triggered once "X" is removed.

Our current implementation, from what I quickly tested, works the same way as Slack's. So I consider this correct.

However, there's a chance that TextWatcher should indeed be improved. For instance, if you a content like this in GH Writer:

```
Text[]
#123
#234
Text
```

And start to navigate through this content with the up/down arrows, you'll be constantly triggering mentions, blocking the navigation. GH's native editor does not have this shortcoming. But... Slack's has if you have content with unfinished mentions:

```
Hello, @re[]
Hello, @re
```

Long story short, I'd be really cautious when changing something, but there's a chance that we should.

There's also a chance that autolink should work differently than mentions.

File: packages/ckeditor5-link/src/autolink.js [hunk 3]
```diff
@@ -223,26 +223,39 @@ export default class AutoLink extends Plugin {
...
-		if ( !this.isEnabled || !isLinkAllowedOnRange( range, model ) || !linkHasProtocol( parsedUrl ) ) {
+		if ( !this.isEnabled || !isLinkAllowedOnRange( range, model ) || !linkHasProtocol( fullUrl ) || linkIsAlreadySet( range ) ) {
 			return;
 		}
 
+		this._persistAutoLink( fullUrl, range );
+	}
+
+	/**
+	 * Enqueue autolink changes in model.
```
Comment (Reinmar) [inline packages/ckeditor5-link/src/autolink.js:246]:
```suggestion
	 * Enqueue autolink changes in the model.
```

**Comment by scofalik** [comment]
I forgot that `TextWatcher` has uses outside of autoformatting text.

> There's also a chance that autolink should work differently than mentions.

Yes. But having mentions in mind, it seems that it is better to add modifications in auto-formatting features (autoformat, autolink, etc.) if needed. We could also manage this behavior through an option parameter when creating `TextWatcher` instance (to what it should react: typing, deleting, selection change).

But this is outside of the scope of this PR. I will create a new issue for this discussion.
